File: rust/src/BTPostOrder.ts
function walk(curr: BinaryNode<number> | null, path: number[]): void {
    if (!curr) {
        return;
    }

    // recurse
    // pre
    
    // recurse
    walk(curr.left, path);
    walk(curr.right, path);
    path.push(curr.value);
    // post, not needed
}

export default function post_order_search(head: BinaryNode<number>): number[] {
    const path: number[] = [];
    walk(head, path);
    return path;
}

File: rust/src/bt_post_order.rs
pub struct BinaryNode<T> {
    pub value: T,
    pub left: Option<Box<BinaryNode<T>>>,
    pub right: Option<Box<BinaryNode<T>>>,
}

fn walk<T>(curr: Option<&Box<BinaryNode<T>>>, path: &mut Vec<T>)
where
    T: Copy,
{
    if let Some(node) = curr {
        walk(node.left.as_ref(), path);
        walk(node.right.as_ref(), path);
        path.push(node.value);
    }
}

pub fn post_order_search(head: &BinaryNode<i32>) -> Vec<i32> {
    let mut path: Vec<i32> = Vec::new();
    walk(head.left.as_ref().or(head.right.as_ref()), &mut path);
    path.push(head.value);
    path
}
